kinisi/parser.py

```diff
@@ -347,3 +371,106 @@ def get_indices(
+
+
+    :return: Tuple containing: Tuple containing: fractional coordinates for centers and framework atoms
+    and Tuple containing: indices for centers used in the calculation 
+    of the diffusion and indices of framework atoms.
+    """
+    drift_indices = []
+    try:
+        indices = indices - 1
+    except:
+        raise ValueError('Molecules must be of same length')
+
+    n_molecules = indices.shape[0]
+
+    # Removed method for framework_indices
+    for i, site in enumerate(structure):
+        if i not in indices.values:
+            drift_indices.append(i)
+
+    if masses == None:
+        weights = sc.ones_like(indices)
+    elif len(masses.values) != indices.values.shape[1]:
+        raise ValueError('Masses must be the same length as a molecule')
+    else:
+        weights = masses.copy()
+
+    new_s_coords = _calculate_centers_of_mass(coords, weights, indices)
+
+    if coords.dtype == np.float32:
+        # MDAnalysis uses float32, so we need to convert to float32 to avoid concat error
+        new_s_coords = new_s_coords.astype(np.float32)
+
+    new_coords = sc.concat([new_s_coords, coords['atom', drift_indices]], 'atom')
+    new_indices = sc.Variable(dims=['molecule'], values=list(range(n_molecules)))
+    new_drift_indices = sc.Variable(dims=['molecule'],
+                                    values=list(range(n_molecules, n_molecules + len(drift_indices))))
+
+    return new_coords, new_indices, new_drift_indices
+
+
+def _calculate_centers_of_mass(coords: VariableLikeType, weights: VariableLikeType,
+                               indices: VariableLikeType) -> VariableLikeType:
+    """
+    Calculates the weighted molecular centre of mass based on chosen weights and indices as per https://doi.org/10.1080/2151237X.2008.10129266
+    The method involves projection of the each coordinate onto a circle to allow for efficient COM calculation
+    
+     :param coords: array of coordinates
+     :param weights: 1D array of weights of elements within molecule
+     :param indices: N by M dimensional array of indices of N molecules of M atoms
+
+     :return: Array containing coordinates of centres of mass of molecules
+    """
+    s_coords = sc.fold(coords['atom', indices.values.flatten()], 'atom', dims=indices.dims, shape=indices.shape)
+    theta = s_coords * (2 * np.pi * (sc.units.rad / sc.units.angstrom))
+    xi = sc.cos(theta)
+    zeta = sc.sin(theta)
+    # This allows the dimensions of the indices to be any word, paired with 'atom'.
+    dims_id = [i for i in indices.dims if i != 'atom'][0]
+    xi_bar = (weights * xi).sum(dim=dims_id) / weights.sum(dim=dims_id)
+    zeta_bar = (weights * zeta).sum(dim=dims_id) / weights.sum(dim=dims_id)
+    theta_bar = sc.atan2(y=-zeta_bar, x=-xi_bar) + np.pi * sc.units.rad
+    new_s_coords = theta_bar / (2 * np.pi * (sc.units.rad / sc.units.angstrom))
```

Comment:
Should this be the `distance_unit` too then?

Reply:
yes

Reply:
Can you have a look to see if there is anywhere else that needs changed too?